core/src/main/java/org/infinispan/container/offheap/ExpandingOffHeapConcurrentMap.java

```diff
@@ -0,0 +1,1131 @@
+package org.infinispan.container.offheap;
+
+import java.lang.invoke.MethodHandles;
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.PrimitiveIterator;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
+import java.util.function.BiFunction;
+import java.util.function.LongConsumer;
+import java.util.stream.LongStream;
+
+import org.infinispan.commons.api.Lifecycle;
+import org.infinispan.commons.marshall.WrappedBytes;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.commons.util.IntSets;
+import org.infinispan.commons.util.IteratorMapper;
+import org.infinispan.commons.util.PeekableMap;
+import org.infinispan.commons.util.ProcessorInfo;
+import org.infinispan.commons.util.Util;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * @author wburns
+ * @since 9.4
+ */
+public class ExpandingOffHeapConcurrentMap implements ConcurrentMap<WrappedBytes, InternalCacheEntry<WrappedBytes, WrappedBytes>>,
+      PeekableMap<WrappedBytes, InternalCacheEntry<WrappedBytes, WrappedBytes>>, Lifecycle {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass());
+   private static final boolean trace = log.isTraceEnabled();
+
+   // Use locks based on numCPUs - but only do up to 1024
+   private final static int LOCK_COUNT = Math.min(Util.findNextHighestPowerOfTwo(ProcessorInfo.availableProcessors()) << 1, 1024);
```

Comment:
Yeah, I thought about sharing the same StripedLock instance between all of the maps (since we have multiple, one for each segment). I didn't think of trying at the level of the CacheManager, but that should be doable as well. This would only be shared by off heap maps obviously. I am not sure if this would cause too much contention between the various maps though for "longer" operations such as resize.

Reply:
Ah yes, the operations that need `lockAll()` might be a problem, especially if we share across caches. I'm crossing my fingers that we'll be able to move the resize allocation under a different lock and replace memoryLookup+oldMemoryLookup+pendingBlocks with memoryLookups[]+oldMemoryLookup so we don't need to hold all the locks at any point of the resize.

Reply:
I am not sure we can move it under a different lock. The reader and writer operations have to have some way to make sure they are reading from the correct lookup and that it isn't changed in the middle of the operation, implying they would have to hold some sort of shared lock with resizing.

Maybe if you can clarify what you are thinking a bit more.

Reply:
Let me see if I can find some time to experiment with this and show you some code. Maybe after we integrate this ;)